### PR TITLE
Implement type conversion/codecs

### LINKED ISSUE #15: Implement type conversion/codecs

### PR DESCRIPTION
**Pull request by dmitry-s**
add codecs to support conversion to and from spanner-specific data types 
fixes #15

### CODE REVIEW

**Comment by elefeint** [comment]
@dmitry-s Could you check why the build is failing?

File: src/test/java/com/google/cloud/spanner/r2dbc/codecs/DefaultCodecsTest.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import com.google.cloud.spanner.Value;
+import java.util.Arrays;
+import java.util.Collection;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
```
Comment (elefeint) [inline src/test/java/com/google/cloud/spanner/r2dbc/codecs/DefaultCodecsTest.java]: Awesome test! Could you also add another test for a few negative scenarios (where the codec is insufficient and would fail to convert).

File: src/main/java/com/google/cloud/spanner/r2dbc/codecs/AbstractCodec.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    public boolean canEncodeNull(Type type) {
+        Assert.requireNonNull(type, "type must not be null");
+
+        return doCanDecode(type);
+    }
+
+    @Nullable
+    @Override
+    public T decode(Struct row, int index, Class<? extends T> type) {
```
Comment (elefeint) [inline src/main/java/com/google/cloud/spanner/r2dbc/codecs/AbstractCodec.java]: Would we call it on (row + column-index) value? I would hope to extract the column before passing it to the codec. 
Column identifiers can be numbers or strings in Row.get(), so letting Row extract the value would save potential duplicate work.

File: src/main/java/com/google/cloud/spanner/r2dbc/codecs/Codec.java [hunk 3]
```diff
@@ -16,26 +16,25 @@
...
 package com.google.cloud.spanner.r2dbc.codecs;
 
-import com.google.cloud.spanner.Struct;
-import com.google.cloud.spanner.Type;
-import com.google.cloud.spanner.Value;
+import com.google.protobuf.Value;
+import com.google.spanner.v1.Type;
 import reactor.util.annotation.Nullable;
 
 interface Codec<T> {
```
Comment (dzou) [inline src/main/java/com/google/cloud/spanner/r2dbc/codecs/Codec.java]: Add javadoc comments for interface methods?

File: src/main/java/com/google/cloud/spanner/r2dbc/codecs/Codec.java [hunk 4]
```diff
@@ -16,26 +16,25 @@
...
 
-    boolean canDecode(Type dataType, Class<?> type);
+  boolean canDecode(Type dataType, Class<?> type);
```
Comment (dzou) [inline src/main/java/com/google/cloud/spanner/r2dbc/codecs/Codec.java]: Add javadoc comments for interface methods?
